### PR TITLE
fix: wire up scaffolded indexHtml to dev servers

### PR DESCRIPTION
**Pull request by ZachJW34**
- Closes UNIFY-1004
### User facing changelog
Renames `indexHtml` property for dev-servers to `indexHtmlFile` and wires up project setup to link to generated `component-index.html` file.

### Additional details
This update adds the `indexHtmlFile` option for both webpack and vite dev-servers (and associated plugins). The majority of the changes are system tests added.

These changes will not be breaking as `indexHtml` was only added on `10.0-release`. I added back in the `template` option to the `@cypress/webpack-dev-server` as this might be used by users. Since each dev-server contains a fallback template, the `indexHtmlFile` option is not required but it will be generated for all new projects that are setup via the launchpad.

To be able to system test these changes, I had to change the `@cypress/mount-utils` dependency from `0.0.0-development` to `*` in `@cypress/react`. It's not ideal, but the system-tests node install will fail otherwise since the `0.0.0-development` link doesn't work for system test projects. I explored ways of fixing this similar to how the system-tests node install will replace linked package dependencies with `file:some-path`, but it was adding a lot more complexity so I decided on this route. `*` dependencies will get replaced with the current version when running `npm-release.js` so it will not affect the final dependencies of the published packages.

### How has the user experience changed?

https://user-images.githubusercontent.com/25158820/156443338-fd73892f-4db0-420d-9a33-31445387ca91.mov

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?
…ex-html

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -67,18 +67,21 @@ packages/data-context/test/unit/codegen/files
 # community templates we test against, no need to lint
 system-tests/projects/create-react-app-configured/**/*
 system-tests/projects/create-react-app-unconfigured/**/*
+system-tests/projects/create-react-app-custom-index-html
```
Comment (lmiller1990) [inline .eslintignore:70]: Not needed now but we could consider grouping all these community template projects w/ a common prefix like `template-` so it's easier to 1) ignore them for linting and 2) group them all together.

File: npm/react/plugins/craco/index.js [hunk 2]
```diff
@@ -2,13 +2,13 @@ const { startDevServer } = require('@cypress/webpack-dev-server')
 const { createWebpackDevConfig } = require('@craco/craco')
 const { getLegacyDevServer } = require('../utils/legacy-setup-dev-server')
 
-function devServer (cypressDevServerConfig, cracoConfig, indexHtml) {
+function devServer (cypressDevServerConfig, cracoConfig, indexHtmlFile) {
```
Comment (lmiller1990) [inline npm/react/plugins/craco/index.js:5]: This will be a breaking change if someone is using `indexHtml`, won't it?

File: packages/data-context/src/actions/WizardActions.ts [hunk 3]
```diff
@@ -313,8 +313,10 @@ export class WizardActions {
       bodyModifier,
     })
 
+    const relativeComponentIndexHtmlPath = path.join('cypress', 'support', 'component-index.html')
+
     return this.scaffoldFile(
-      path.join(this.projectRoot, 'cypress', 'component', 'index.html'),
+      path.join(this.projectRoot, relativeComponentIndexHtmlPath),
```
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:319]: Could just include `this.projectRoot` in the `path.join` on line 316? No need to do `path.join` twice.

File: packages/scaffold-config/src/frameworks.ts [hunk 4]
```diff
@@ -43,7 +43,10 @@ export const FRONTEND_FRAMEWORKS = [
       module.exports = {
         component: {
           devServer,
-        }
+          devServerConfig: {
```
Comment (lmiller1990) [inline packages/scaffold-config/src/frameworks.ts:46]: We definitely need a better system for all these configs other than copy paste. That said, we had string interpolation before and it was not ideal, so I'm okay with this for now. Let's try and think of something more scalable and less boilerplate-y in the future, though.

File: system-tests/projects/create-react-app-custom-index-html/.gitignore [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Comment (lmiller1990) [inline system-tests/projects/create-react-app-custom-index-html/.gitignore:1]: I think we should remove all the nested `.gitignore` files in the community templates, they don't add much value.

File: system-tests/projects/create-react-app-custom-index-html/src/App.cy.jsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { mount } from '@cypress/react'
+import React from 'react'
+import App from './App'
+
+it('works', () => {
+  mount(<App />)
+  cy.contains('Learn React')
+  cy.get('body').should('have.css', 'background-color', 'rgb(255, 0, 0)')
```
Comment (lmiller1990) [inline system-tests/projects/create-react-app-custom-index-html/src/App.cy.jsx:8]: Nice, good idea to add this to make sure it's using the right index.html.

File: system-tests/projects/vue3-vite-ts-custom-index-html/.gitignore [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+*.local
+
+# Editor directories and files
+.vscode/*
+!.vscode/extensions.json
+.idea
+.DS_Store
+*.suo
+*.ntvs*
+*.njsproj
```
Comment (lmiller1990) [inline system-tests/projects/vue3-vite-ts-custom-index-html/.gitignore:22]: Let's remove this

File: system-tests/projects/vueclivue3-custom-index-html/package.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  "private": true,
+  "scripts": {
+    "serve": "vue-cli-service serve",
+    "build": "vue-cli-service build"
+  },
+  "dependencies": {
+    "vue": "^3.0.0"
+  },
+  "devDependencies": {
+    "@cypress/vue": "^3.0.0",
```
Comment (lmiller1990) [inline system-tests/projects/vueclivue3-custom-index-html/package.json:13]: Should this be using `file:../../../npm/vue`?

File: system-tests/projects/yarn-v3.1.1-pnp/yarn.lock [hunk 9]
```diff
@@ -3,12 +3,12 @@
 
 __metadata:
   version: 5
-  cacheKey: 7
+  cacheKey: 8
```
Comment (lmiller1990) [inline system-tests/projects/yarn-v3.1.1-pnp/yarn.lock:6]: Was this updated intentionally? Seems unrelated.

File: system-tests/test/component_testing_index_html_spec.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import systemTests from '../lib/system-tests'
+
+describe('component testing projects', function () {
+  systemTests.setup()
+
+  systemTests.it('create-react-app-custom-index-html', {
+    project: 'create-react-app-custom-index-html',
+    testingType: 'component',
+    spec: 'src/App.cy.jsx',
```
Comment (lmiller1990) [inline system-tests/test/component_testing_index_html_spec.ts:9]: More a question for @flotwig, is there some way to say "run all the specs in the project" for a system test project, instead of specifying one here? I tried a few things but didn't get it to work.

File: system-tests/test/component_testing_index_html_spec.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    browser: 'chrome',
+    expectedExitCode: 0,
+  })
+
+  systemTests.it('vueclivue3-custom-index-html', {
+    project: 'vueclivue3-custom-index-html',
+    testingType: 'component',
```
Comment (lmiller1990) [inline system-tests/test/component_testing_index_html_spec.ts:16]: Definitely like these system tests, but we also should keep an eye on how much time they add to CI (if they do) and be conscious of it.

File: yarn.lock [hunk 12]
```diff
@@ -2628,6 +2628,11 @@
   dependencies:
     mocha ">=1.13.0"
 
+"@cypress/mount-utils@*":
```
Comment (lmiller1990) [inline yarn.lock:2631]: I think this update is probably unintentional/not needed?

**Review by lmiller1990 — COMMENTED** [review]
Left a few comments. I wonder what it would take to apply the same logic we apply to get `file:.../` in `package.json` to work for system tests to also apply for dependencies of `npm/react` that use `file:.../`, primarily `npm/mount-utils`. :thinking:

**Comment by lmiller1990** [comment]
Great PR, especially the description and video :100:

File: packages/scaffold-config/src/frameworks.ts (hunk 4 above)
Comment (elevatebart) [reply packages/scaffold-config/src/frameworks.ts:46]: I agree. 
The usual best practice solution is to as little code in those files as possible.
We should add all this in the defaults. This way we can improve the defaults as we improve the rest. 

Another argument to not scaffold this line or make it a required file is that... well, we have no use for it internally do we?

File: packages/scaffold-config/src/frameworks.ts [hunk 13]
```diff
@@ -87,7 +93,10 @@ export const FRONTEND_FRAMEWORKS = [
...
+            indexHtmlFile: 'cypress/support/component-index.html',
```
Comment (elylucas) [inline packages/scaffold-config/src/frameworks.ts:97]: If this is the default value, I think I’d prefer it to not be in the template. I’ve been told that whatever we can do (in reason) to minimize the amount of setup in the config file should be done

File: scripts/npm-release.js [hunk 14]
```diff
@@ -101,7 +101,7 @@ const getPackageVersions = async (packages) => {
 }
 
 // updates a public package's package.json
-// replaces any local dependencies that have a * version
+// replaces any local dependencies that have a * or file: version
```
Comment (elevatebart) [inline scripts/npm-release.js:104]: To revert

**Review by elevatebart — APPROVED** [review]

File: npm/react/plugins/craco/index.js (hunk 2 above)
Comment (ZachJW34) [reply npm/react/plugins/craco/index.js:5]: `indexHtml` was only added on `10.0-release` so this update would not be breaking since it was never released.

File: .eslintignore (hunk 1 above)
Comment (ZachJW34) [reply .eslintignore:70]: Can we just lint ignore all system-test projects?

File: system-tests/test/component_testing_index_html_spec.ts (hunk 10 above)
Comment (flotwig) [reply system-tests/test/component_testing_index_html_spec.ts:9]: Omitting the `spec` parameter should do it. Or `*`, `**/*`... it's passed straight to `--spec`, so anything like that should work.

File: system-tests/test/component_testing_index_html_spec.ts (hunk 11 above)
Comment (flotwig) [reply system-tests/test/component_testing_index_html_spec.ts:16]: May be time to up the `parallelization` on these jobs in `circle.yml`.

**Review by elevatebart — APPROVED** [review]

File: .eslintignore (hunk 1 above)
Comment (lmiller1990) [reply .eslintignore:71]: Not sure, maybe, but a discussion and task for another PR, this is okay for now.

File: npm/react/plugins/craco/index.js (hunk 2 above)
Comment (lmiller1990) [reply npm/react/plugins/craco/index.js:5]: Oh, I thought this was in develop. All good.

**Review by lmiller1990 — APPROVED** [review]
